### PR TITLE
feat: ERPNext Open Banking integration app

### PR DESCRIPTION
**Pull request by john-frandsen**
## ERPNext / Frappe app for open-banking.io

Syncs bank transactions from open-banking.io into ERPNext's `Bank Transaction` doctype — ready for the Bank Reconciliation Tool.

### Why
ERPNext users on [discuss.frappe.io](https://discuss.frappe.io) are paying **$1,000–2,000/month for Plaid** bank feeds. open-banking.io provides the same PSD2 account data at **~€3/month** with no eIDAS certificate requirement. This app makes that accessible to the ERPNext ecosystem.

### What's included
- **3 doctypes:** Open Banking Settings (credentials), Open Banking Connection (OBI account → ERPNext Bank Account mapping), Open Banking Sync Log (per-sync audit)
- **Vendored zero-knowledge decryption:** ECDH P-256 → HKDF-SHA256 → AES-256-GCM (same scheme as the Python SDK + n8n node)
- **Sync engine:** fetches transactions from OBI → triggers online sync → paginates → maps to ERPNext Bank Transaction fields → deduplicates by transaction_id
- **Scheduler:** runs every 6 hours automatically
- **UI:** "Sync from Open Banking" button on Bank Account forms, Test Connection on Settings
- **15 unit tests** (envelope roundtrip + transaction mapping edge cases)

### Structure
Follows the existing monorepo pattern — `erpnext/` sits alongside `n8n/` and the language SDKs:

```
erpnext/
├── erpnext_open_banking/
│   ├── hooks.py
│   ├── modules.txt
│   ├── erpnext_open_banking/          # module
│   │   ├── doctype/                   # 3 doctypes
│   │   ├── utils/                     # envelope, client, mapper, connector
│   │   └── config/
│   └── public/js/bank_account.js      # Sync button
├── tests/                             # 15 tests, no Frappe dependency needed
├── pyproject.toml
└── README.md
```

### Zero external dependencies
Uses only `cryptography` + `requests` — both already shipped by Frappe. No pip installs beyond what ERPNext provides.

### Disclosure
I'm the maintainer of open-banking.io. This is an official integration app, MIT-licensed, same as the n8n node and language SDKs in this repo.

## Summary by CodeRabbit

* **New Features**
  * Added Open Banking integration setup, connection tracking, transaction sync, and sync logs in ERPNext.
  * Bank Account and settings screens now include actions to test connectivity and trigger syncs.
  * Scheduled syncing can run automatically every 6 hours when enabled.

* **Documentation**
  * Added setup, usage, and security guidance for the Open Banking app.

* **Bug Fixes**
  * Improved transaction mapping and sync handling to better avoid duplicates and surface errors clearly.

### CODE REVIEW

File: .github/workflows/erpnext-tests.yml [hunk 1]
```diff
@@ -0,0 +1,30 @@
...
+
+on:
+  push:
+    paths:
+      - 'erpnext/**'
+      - '.github/workflows/erpnext-tests.yml'
+  pull_request:
+    paths:
+      - 'erpnext/**'
+      - '.github/workflows/erpnext-tests.yml'
```
[the remark this answers is not in the document]
Comment (nicolaj-hartmann) [reply .github/workflows/erpnext-tests.yml]: Fixed in faf7885 by removing this workflow entirely — the erpnext tests now run as a job in ci.yml, gated by the changed-packages filter and aggregated by ci-gate like every other package in this repo.

File: .github/workflows/erpnext-tests.yml [hunk 2]
```diff
@@ -0,0 +1,30 @@
...
+    strategy:
+      fail-fast: false
+      matrix:
+        python-version: ["3.9", "3.11", "3.13"]
+    defaults:
+      run:
+        working-directory: erpnext
+    steps:
+      - uses: actions/checkout@v4
+      - uses: actions/setup-python@v5
```
[the remark this answers is not in the document]
Comment (nicolaj-hartmann) [reply .github/workflows/erpnext-tests.yml]: Fixed in faf7885 — the standalone workflow is gone; the replacement job in ci.yml uses the repo-standard SHA-pinned actions/checkout and actions/setup-python.

File: erpnext/erpnext_open_banking/erpnext_open_banking/doctype/open_banking_connection/open_banking_connection.json [hunk 3]
```diff
@@ -0,0 +1,166 @@
...
+  "valid_until"
+ ],
+ "fields": [
+  {
+   "fieldname": "obi_account_id",
+   "fieldtype": "Data",
+   "label": "OBI Account ID",
+   "reqd": 1,
+   "read_only": 1
+  },
```
[the remark this answers is not in the document]
Comment (nicolaj-hartmann) [reply erpnext/erpnext_open_banking/erpnext_open_banking/doctype/open_banking_connection/open_banking_connection.json:34]: Fixed in faf7885 — added `unique: 1` on `obi_account_id`.

File: erpnext/erpnext_open_banking/erpnext_open_banking/doctype/open_banking_settings/open_banking_settings.json [hunk 4]
```diff
@@ -0,0 +1,104 @@
...
+  },
+  {
+   "fieldname": "credentials_bundle",
+   "fieldtype": "Code",
+   "label": "Credentials Bundle (JSON)",
+   "options": "JSON",
+   "reqd": 1,
+   "description": "Paste the credentials bundle JSON you exported from open-banking.io. It contains apiBaseUrl, apiKey and encryptionKey.privateKey. The private key never leaves this server — it is used only to decrypt data locally.",
+   "length": 10000
+  },
```
[the remark this answers is not in the document]
Comment (nicolaj-hartmann) [reply erpnext/erpnext_open_banking/erpnext_open_banking/doctype/open_banking_settings/open_banking_settings.json:31]: Fixed in faf7885 — `credentials_bundle` is now a `Password` field (stored encrypted, read server-side via `get_password`), and print/email permissions were dropped from the Settings doctype.

File: erpnext/erpnext_open_banking/erpnext_open_banking/doctype/open_banking_settings/open_banking_settings.py [hunk 5]
```diff
@@ -0,0 +1,38 @@
...
+        connections = client.get_connections()
+        client.close()
+        return {
+            "success": True,
+            "message": _(
+                "Connected successfully. {0} bank connection(s) found."
+            ).format(len(connections)),
+        }
+    except Exception as exc:
+        return {"success": False, "message": str(exc)}
```
[the remark this answers is not in the document]
Comment (nicolaj-hartmann) [reply erpnext/erpnext_open_banking/erpnext_open_banking/doctype/open_banking_settings/open_banking_settings.py]: Fixed in faf7885 — `test_connection` now calls `frappe.only_for(("Accounts Manager", "System Manager"))` and is restricted to POST. Worth noting: the analogous check in `sync_now` used `frappe.has_role`, which does not exist in Frappe — it would have raised AttributeError for every non-Administrator caller. Both now use `frappe.only_for`.

File: erpnext/erpnext_open_banking/erpnext_open_banking/doctype/open_banking_sync_log/open_banking_sync_log.json [hunk 6]
```diff
@@ -0,0 +1,139 @@
...
+   "label": "Errors"
+  },
+  {
+   "fieldname": "error_detail",
+   "fieldtype": "Small Text",
+   "label": "Error Detail",
+   "read_only": 1
+  }
+ ],
+ "in_create": 1,
```
[the remark this answers is not in the document]
Comment (nicolaj-hartmann) [reply erpnext/erpnext_open_banking/erpnext_open_banking/doctype/open_banking_sync_log/open_banking_sync_log.json:105]: Declining — `in_create` is a valid DocType-level property in Frappe (it's defined on the DocType doctype itself, see frappe/core/doctype/doctype/doctype.json); at the root it prevents manual record creation, which is exactly the intent for a log doctype.

File: erpnext/erpnext_open_banking/erpnext_open_banking/doctype/open_banking_sync_log/open_banking_sync_log.json [hunk 7]
```diff
@@ -0,0 +1,139 @@
...
+   "email": 1,
+   "export": 1,
+   "print": 1,
+   "read": 1,
+   "report": 1,
+   "role": "Accounts Manager",
+   "share": 1,
+   "write": 1
+  }
+ ],
```
[the remark this answers is not in the document]
Comment (nicolaj-hartmann) [reply erpnext/erpnext_open_banking/erpnext_open_banking/doctype/open_banking_sync_log/open_banking_sync_log.json:130]: Fixed in faf7885 — Accounts Manager is now limited to read/report/export/print/email on the sync log; create/write/delete remain System Manager only.

File: erpnext/erpnext_open_banking/erpnext_open_banking/utils/client.py [hunk 8]
```diff
@@ -0,0 +1,285 @@
...
+            "debtor_bban": d.get("debtorBban"),
+            "debtor_agent_bic": d.get("debtorAgentBic"),
+            "remittance_information": d.get("remittanceInformation"),
+            "note": d.get("note"),
+            "reference_number": d.get("referenceNumber"),
+            "exchange_rate": d.get("exchangeRate"),
+            "merchant_category_code": d.get("merchantCategoryCode"),
+            "balance_after_transaction": d.get("balanceAfter"),
+            "balance_after_currency": d.get("balanceAfterCurrency"),
+        }
```
[the remark this answers is not in the document]
Comment (nicolaj-hartmann) [reply erpnext/erpnext_open_banking/erpnext_open_banking/utils/client.py:292]: Fixed in faf7885 — added `_parse_optional_decimal` which parses via `Decimal(str(value))` and preserves `None` for absent values; applied to `exchange_rate` and `balance_after_transaction`.

File: erpnext/erpnext_open_banking/erpnext_open_banking/utils/connector.py [hunk 9]
```diff
@@ -0,0 +1,245 @@
...
+def get_client(settings: dict[str, Any] | None = None) -> OpenBankingClient:
+    """Builds an OpenBankingClient from the Open Banking Settings doctype."""
+    if settings is None:
+        settings = frappe.get_doc("Open Banking Settings").as_dict()
+
+    bundle_raw = settings.get("credentials_bundle", "")
+    if not bundle_raw:
+        frappe.throw(_("Please paste your open-banking.io credentials bundle in Open Banking Settings."))
+
+    return OpenBankingClient.from_credentials(bundle_raw)
```
[the remark this answers is not in the document]
Comment (nicolaj-hartmann) [reply erpnext/erpnext_open_banking/erpnext_open_banking/utils/connector.py]: Fixed in faf7885 — `from_credentials` gained a `base_url_override` keyword and `get_client` passes `api_base_url_override` through; `test_connection` now reuses `get_client` so both paths honour the override.

File: erpnext/erpnext_open_banking/erpnext_open_banking/utils/connector.py [hunk 10]
```diff
@@ -0,0 +1,245 @@
...
+
+    Returns ``{"created": int, "skipped": int, "total": int, "errors": list}``.
+    """
+    conn = frappe.get_doc("Open Banking Connection", connection_name)
+    settings = frappe.get_doc("Open Banking Settings").as_dict()
+
+    if conn.status != "Active":
+        return {"created": 0, "skipped": 0, "total": 0, "errors": ["Connection is not active"]}
+
+    company = settings.get("default_company") or conn.company
```
[the remark this answers is not in the document]
Comment (nicolaj-hartmann) [reply erpnext/erpnext_open_banking/erpnext_open_banking/utils/connector.py]: Fixed in faf7885 — precedence flipped to `conn.company or settings.get("default_company")`.

File: erpnext/erpnext_open_banking/erpnext_open_banking/utils/connector.py [hunk 11]
```diff
@@ -0,0 +1,245 @@
...
+            "accounts_fetched": 1,
+            "transactions_created": created,
+            "transactions_skipped": skipped,
+            "total_available": total,
+            "status": "Completed" if not errors else "Completed with errors",
+            "error_detail": "\n".join(errors) if errors else "",
+        }
+    )
+
+    return {"created": created, "skipped": skipped, "total": total, "errors": errors}
```
[the remark this answers is not in the document]
Comment (nicolaj-hartmann) [reply erpnext/erpnext_open_banking/erpnext_open_banking/utils/connector.py:213]: Fixed in faf7885 — `sync_connection` now wraps the sync in a per-connection `filelock` (frappe.utils.synchronization) with a 5s timeout; a concurrent run returns a friendly "already running" result instead of double-inserting.

File: erpnext/erpnext_open_banking/erpnext_open_banking/utils/connector.py [hunk 12]
```diff
@@ -0,0 +1,245 @@
...
+
+        # Batch-fetch existing transaction_ids for this bank account (avoids
+        # N+1 queries — one SELECT instead of one per transaction).
+        existing_ids = set(
+            frappe.get_all(
+                "Bank Transaction",
+                filters={"bank_account": bank_account},
+                pluck="transaction_id",
+            )
+        )
```
[the remark this answers is not in the document]
Comment (nicolaj-hartmann) [reply erpnext/erpnext_open_banking/erpnext_open_banking/utils/connector.py]: Fixed in faf7885 — the dedup query now filters `transaction_id IN (fetched ids)`, so it's bounded by the fetched batch and exact (no date-window heuristic that could miss cross-window duplicates).

File: erpnext/erpnext_open_banking/erpnext_open_banking/utils/connector.py [hunk 13]
```diff
@@ -0,0 +1,245 @@
...
+
+
+@frappe.whitelist()
+def sync_now(bank_account: str | None = None) -> dict[str, Any]:
+    """Whitelisted method for the 'Sync Now' button on the Bank Account form.
+
+    Only Accounts Managers (or System Managers) may trigger a sync.
+    """
+    if not frappe.has_role(["Accounts Manager", "System Manager"]):
+        frappe.throw(_("Only Accounts Managers can trigger bank sync."), frappe.PermissionError)
```
[the remark this answers is not in the document]
Comment (nicolaj-hartmann) [reply erpnext/erpnext_open_banking/erpnext_open_banking/utils/connector.py]: Fixed in faf7885 — `@frappe.whitelist(methods=["POST"])`. The form button is unaffected since `frappe.call` defaults to POST.

File: erpnext/erpnext_open_banking/public/js/bank_account.js [hunk 14]
```diff
@@ -0,0 +1,46 @@
+// Bank Account form — adds "Sync Now" button that triggers open-banking.io sync
+
+frappe.ui.form.on("Bank Account", {
+    refresh(frm) {
+        // Only show button if an Open Banking Connection is linked to this Bank Account
+        frappe.db
+            .get_value("Open Banking Connection", { bank_account: frm.doc.name }, "name")
+            .then((r) => {
+                if (r && r.message && r.message.name) {
```
[the remark this answers is not in the document]
Comment (nicolaj-hartmann) [reply erpnext/erpnext_open_banking/public/js/bank_account.js:14]: Fixed in faf7885 — the refresh handler now returns early for users without Accounts Manager/System Manager and the `get_value` promise has a `.catch()` fallback.

File: erpnext/tests/test_envelope.py [hunk 15]
```diff
@@ -0,0 +1,129 @@
...
+        bad = b"\x02" + b"\x00" * 100
+        with pytest.raises(ValueError, match="Invalid or unsupported envelope"):
+            decrypt(loaded, bad)
+
+    def test_decrypt_too_short(self):
+        """decrypt rejects an envelope shorter than the minimum header."""
+        priv, pkcs8_b64 = _generate_keypair()
+        loaded = load_private_key(pkcs8_b64)
+        with pytest.raises(ValueError, match="Invalid or unsupported envelope"):
+            decrypt(loaded, b"\x01\x00\x01")
```
[the remark this answers is not in the document]
Comment (nicolaj-hartmann) [reply erpnext/tests/test_envelope.py:129]: Fixed in faf7885 — added tests for tampered ciphertext, tampered tag, and wrong recipient key (all must raise InvalidTag).

File: .github/workflows/ci.yml [hunk 16]
```diff
@@ -242,12 +243,30 @@ jobs:
...
+    if: needs.changes.outputs.erpnext == 'true'
+    runs-on: ubuntu-latest
+    strategy:
+      matrix:
+        python-version: ['3.9', '3.11', '3.13']
+    defaults:
+      run:
+        working-directory: erpnext
+    steps:
+      - uses: actions/checkout@9c091bb21b7c1c1d1991bb908d89e4e9dddfe3e0 # v7.0.0
```
[the remark this answers is not in the document]
Comment (nicolaj-hartmann) [reply .github/workflows/ci.yml:257]: Declining for this PR — all 13 jobs in ci.yml share the same checkout pattern; flipping persist-credentials repo-wide deserves its own focused PR (same note as on #96).

File: .github/workflows/ci.yml [hunk 17]
```diff
@@ -242,12 +243,30 @@ jobs:
...
+      - uses: actions/setup-python@ece7cb06caefa5fff74198d8649806c4678c61a1 # v6.3.0
+        with:
+          python-version: ${{ matrix.python-version }}
+      - run: pip install pytest cryptography requests
```
[the remark this answers is not in the document]
Comment (nicolaj-hartmann) [reply .github/workflows/ci.yml:261]: Declining — the app's runtime deps are whatever Frappe ships (cryptography/requests), so the tests deliberately track current releases to catch breakage early, matching how the other language jobs test against current toolchains. There's no lockfile to pin against, and hard pins here would just rot.

**Review by nicolaj-hartmann — APPROVED** [review]
Reviewed the full diff plus the CodeRabbit findings. Pushed faf7885 addressing the substantive ones: encrypted Password storage for the credentials bundle, working role checks (frappe.has_role does not exist — replaced with frappe.only_for), POST-only whitelisted methods, per-connection sync lock, company-precedence fix, base-URL override wiring, and CI integration behind changed-packages/ci-gate. Tests pass on 3.9/3.11/3.13. LGTM.
